File: 2024/06/cpp/06.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long      ll;
typedef vector<int>    vi;
typedef vector<bool>   vb;
typedef pair<int, int> ii;

const int INF = 0x3f3f3f3f;
const ll  INFLL = 0x3f3f3f3f3f3f3f3fll;

#define each(x, s)  for (auto& x : s)
#define loop(x)     for (int i = 0;i < x;i++)
#define vloop(v, x) for (int v = 0;v < x;v++)
#define iter(a)     a.begin(), a.end()
#define riter(a)    a.rbegin(), a.rend()
#define endl        "\n"

int task1(vector<string> input) {
  int n = input.size();
  int m = input[0].size();
  int x, y;
  loop(n) vloop(j, m) {
    if (input[i][j] == '^') {
      x = i;
      y = j;
      i = j = INF;
    }
  }
  vector<ii> pos;
  int dir = 0;
  while (true) {
    pos.emplace_back(x, y);
    int nx = x + (dir == 2) - (dir == 0);
    int ny = y + (dir == 1) - (dir == 3);
    if (nx < 0 or nx >= n or ny < 0 or ny >= m) break;
    if (input[nx][ny] == '#') {
      dir = (dir + 1) % 4;
    }
    else {
      x = nx;
      y = ny;
    }
  }
  sort(iter(pos));
  return unique(iter(pos)) - pos.begin();
}

int task2(vector<string> input) {
  int n = input.size();
  int m = input[0].size();
  int x, y;
  loop(n) vloop(j, m) {
    if (input[i][j] == '^') {
      x = i;
      y = j;
      i = j = INF;
    }
  }
  auto check = [&] (int x, int y) {
    vector<vi> vis(n, vi(m));
    int dir = 0;
    while (true) {
      if (vis[x][y] & (1 << dir)) return true;
      vis[x][y] |= 1 << dir;
      int nx = x + (dir == 2) - (dir == 0);
      int ny = y + (dir == 1) - (dir == 3);
      if (nx < 0 or nx >= n or ny < 0 or ny >= m) return false;
      if (input[nx][ny] == '#') dir = (dir + 1) % 4;
      else {
        x = nx;
        y = ny;
      }
    }
    return true;
  };
  int ans = 0;
  loop(n) vloop(j, n) {
    if (input[i][j] == '#') continue;
    input[i][j] = '#';
    ans += check(x, y);
    input[i][j] = '.';
  }
  return ans;
}

int main() {
  ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);
  
  string s;
  vector<string> input;
  while (getline(cin, s)) input.push_back(s);
  cout << "Task 1: " << task1(input) << endl;
  cout << "Task 2: " << task2(input) << endl;
  return 0;
}

File: 2024/06/rust/06.rs
use std::io;

fn task1(input: Vec<Vec<char>>) -> i32 {
    let n = input.len() as i32;
    let m = input[0].len() as i32;
    let mut x = 0;
    let mut y = 0;
    for i in 0..n {
        for j in 0..m {
            if input[i as usize][j as usize] == '^' {
                x = i;
                y = j;
            }
        }
    }
    let mut dir = 0;
    let mut vis = Vec::new();
    loop {
        vis.push((x, y));
        let nx = x + [-1, 0, 1, 0][dir as usize];
        let ny = y + [0, 1, 0, -1][dir as usize];
        if nx < 0 || nx >= n || ny < 0 || ny >= m {
            break;
        }
        if input[nx as usize][ny as usize] == '#' {
            dir = (dir + 1) % 4;
        } else {
            x = nx;
            y = ny;
        }
    }
    // remove duplicates
    vis.sort();
    vis.dedup();
    vis.len() as i32
}

fn task2(mut input: Vec<Vec<char>>) -> i32 {
    let n = input.len() as i32;
    let m = input[0].len() as i32;
    let mut x = 0;
    let mut y = 0;
    for i in 0..n {
        for j in 0..m {
            if input[i as usize][j as usize] == '^' {
                x = i;
                y = j;
            }
        }
    }
    let mut ans = 0;
    for i in 0..n {
        for j in 0..m {
            if input[i as usize][j as usize] == '#' {
                continue;
            }
            input[i as usize][j as usize] = '#';

            let mut vis = vec![vec![0 as i32; m as usize]; n as usize];
            let mut dir = 0;
            let mut a = x;
            let mut b = y;
            loop {
                if vis[a as usize][b as usize] & (1 << dir) != 0 {
                    ans += 1;
                    break;
                }
                vis[a as usize][b as usize] |= 1 << dir;
                let nx = a + [-1, 0, 1, 0][dir];
                let ny = b + [0, 1, 0, -1][dir];
                if nx < 0 || nx >= n || ny < 0 || ny >= m {
                    break;
                }
                if input[nx as usize][ny as usize] == '#' {
                    dir = (dir + 1) % 4;
                }
                else {
                    a = nx;
                    b = ny;
                }
            }

            input[i as usize][j as usize] = '.';
        }
    }
    ans
}

fn main() {
    let mut s = String::new();
    let mut input = Vec::new();
    while io::stdin().read_line(&mut s).unwrap() > 0 {
        input.push(s.trim().chars().collect());
        s.clear();
    }
    println!("Task 1: {}", task1(input.clone()));
    println!("Task 2: {}", task2(input.clone()));
}
